idioms/coercion-arguments.md

```diff
@@ -0,0 +1,117 @@
+            'a' | 'e' | 'i' | 'o' | 'u' => {
+                vowel_count += 1;
+                if vowel_count >= 3 {
+                    return true
+                }
+            }
+            _ => vowel_count = 0
+        }
+    }
+    false
+}
+
+fn main() {
+    let ferris = "Ferris".to_string();
+    let curious = "Curious".to_string();
+    println!("{}: {}", ferris, three_vowels(&ferris));
+    println!("{}: {}", curious, three_vowels(&curious));
+
+    // Using a `&String` type in our argument we will find the following example fails:
+    // println!("Ferris: {}", three_vowels("Ferris"));
+    // println!("Curious: {}", three_vowels("Curious"));
+
+}
+```
+
+This works fine because we are passing a `&String` type as a parameter.
+If we comment in the last two lines this example fails because a `&str` type will not coerce to a `&String` type. 
+We can fix this by simply modifying the type for our argument. 
+
+For instance, if we change our function declaration to:
+
+```rust, ignore
+fn three_vowels(word: &str) -> bool {
+```
+
+then both versions will compile and print the same output.
+
+```bash
+Ferris: false
+Curious: true
+```
+
+But wait, that's not all! There is more to this story. 
+It's likely that you may say to yourself: that doesn't matter, I will never be using a `&'static str` as an input anways (as we did when we used `"Ferris"`).
+Even ignoring this special example, you may still find that using `&str` will give you more flexibility than using a `&String`.
+
+Let's now take an example where someone gives us a sentence, and we want to determine if any of the words in the sentence has a word that contains three consecutive vowels.
+We probably should make use of the function we have already defined and simply feed in each word from the sentence.
+
+An example of this could look like this:
+
+```rust
+fn three_vowels(word: &str) -> bool {
+    let mut vowel_count = 0;
+    for c in word.chars() {
+        match c {
+            'a' | 'e' | 'i' | 'o' | 'u' => {
+                vowel_count += 1;
+                if vowel_count >= 3 {
+                    return true
+                }
+            }
+            _ => vowel_count = 0
+        }
+    }
+    false
+}
+
+fn main() {
+    let sentence_string = 
+        "Once upon a time, there was a friendly curious crab named Ferris".to_string();
+    for word in sentence_string.split(' ') {
+        if three_vowels(word) {
+            println!("{} has three consecutive vowels!", word);
+        }
+    }
+}
+```
+
+Running this example using our function declared with an argument type `&str` will yield
+
+```bash
+curious has three consecutive vowels!
+```
+
+However, this example will not run when our function is declared with an argument type `&String`. 
+This is because string slices are a `&str` and not a `&String` which would require an allocation to be converted to `&String` which is not implicit, whereas converting from `String` to `&str` is cheap and implicit.
+
```

Comment:
Missing:
- Discussion
- Advantages
- Disadvantages

Reply:
mmm..Do you mean that this chapter should be restructured to include those?

Because I don't feel like other content should be added. There is the risk that we create repetitions.

Reply:
I felt that this could take out repetition and bundle it under the mentioned points instead of spreading it over the article. 🤔 Just an idea though.

Reply:
I'm happy to land this now and iterate on it later. This document might not be perfect but I think it's good enough, and a rewrite will probably drag this PR out too much.

Reply:
Agreed ;-)